File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        int maxLength = 0;
        int left = 0;

        unordered_map<char, int> charIndex; 

        for (int right = 0; right < n; ++right) 
        {
            char currentChar = s[right];
            if (charIndex.find(currentChar) != charIndex.end()) 
            {
                left = max(left, charIndex[currentChar] + 1);
            }
            charIndex[currentChar] = right;
            int currentLength = right - left + 1;

            maxLength = max(maxLength, currentLength);
        }

        return maxLength;
    }
};
/*
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int i = 0, len = 0;
        string myString = "";

        while (i < s.length()) 
        {
            if (myString.find(s[i]) == string::npos) 
            {
                myString.push_back(s[i]);
                i++;
            }
            else 
            {
                len = max(len, static_cast<int>(myString.length()));
                myString.clear();
            }
        }

        len = max(len, static_cast<int>(myString.length()));

        return len;
    }
};
*/

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        left = 0
        maxLength = 0
        charIndex = {}

        for right in range (n):
            currentChar = s[right]
            if currentChar in charIndex:
                left = max(left,charIndex[currentChar]+1)
            
            charIndex[currentChar] = right
            currentLength = right - left + 1
            maxLength = max(maxLength,currentLength)

        return maxLength